vision/objectdetection/color_detector.go

```diff
@@ -24,14 +26,23 @@ func NewColorDetector(cfg *ColorDetectorConfig) (Detector, error) {
 		return nil, err
 	}
 	hue, _, _ := col.HsvNormal()
-	tol := cfg.Tolerance
+	tol := cfg.HueTolerance
 	if tol > 1.0 || tol < 0.0 {
 		return nil, errors.Errorf("tolerance must be between 0.0 and 1.0. Got %.5f", tol)
 	}
 
+	sat := cfg.SaturationCutoff
+	if sat <= 0.0 {
+		sat = 0.2
+	}
```

Comment:
Why not just error for negative values like we do for `HueTolerance` above?

Reply:
Good point- if a user explicitly enters a negative value, thats an error rather than a call to set a default value.

Reply:
Also, how are we distinguishing between a 0 value set by a user and a 0 value initialized by default? It seems we could either

1. **Not let users set any values to 0.** This is probably the simplest way and would still allow users to set arbitrarily low values, e.g. 0.000001, since it's a float. Though, this behavior seems strange and users might wonder why setting one of these values to zero doesn't give them the expected result.
2. **Make these fields a pointer to a number type.** This would make the struct a bit strange to use since you'd have to dereference the pointer each time you use it.
3. **Encapsulate the struct.** That is, make the fields of the struct private and create a `get_new_color_detector_stuct` type function. I don't think Go allows optional/default parameters though. Maybe we could use variable arguments but we'd have to be careful to not accidentally swap any of the number values. Swapping them wouldn't throw a compile-time error (or any error) and would just silently give the wrong results.

Reply:
I've changed this to error on negatives (and also catch values above 1.0 as errors too), and only treat exactly zero as a default case.

As for the 0==default thing... we do that a lot, all throughout RDK, so it doesn't feel that strange to me. Also, because of how we use json parsers to load our configurations, these always come in as being set, even if omitted, so as far as I know, there's no way to distinguish between "user set parameter explicitly to zero" and "json parser filled a missing field with zero" without changing how we parse configurations.

Reply:
I do think making the data type a pointer is a way of doing this- not setting the value explicitly to zero in the config will make it nil instead. So if a variable == nil it means it should be set to the default value.

Reply:
We wouldn't be able to tell the JSON parser to only fill in a missing value with our default parameter if one doesn't exist? If this is using Go struct tags then I guess that'd be something like `json:"data,omitempty"`. I see your point about it being consistent though so I'm not arguing either way. Just wanted to make sure that we're explicitly deciding that users can not set any of these values to 0.

Reply:
But since we do use omitempty for the sat and val thresholds, those could technically use the pointer method to allow the user to set those fields explicitly to 0. But the threshold i think should never be 0 anyway, so in my mind it's ok that it wont be allowed to be set to 0

Reply:
To walk back what I said earlier, we CAN set a custom UnmarshallJSON() method on the config struct, but it has to be maintained in sync with the config struct itself, and directly deal with all the fields. I started writing one just to see how complex it'd be, and...

```
func (cfg *ColorDetectorConfig) UnmarshalJSON(dat []byte) error {
	var cfgMap map[string]interface{}
	err := json.Unmarshal(b, &cfgMap)
	if err != nil {
		return err
	}

	// Repeat the rest for ALL fields
	rawVal, ok := cfgMap["segment_size_px"]
	if ok {
		cfg.SegmentSize, err = strconv.Atoi(rawVal)
		if err != nil {
			return err
		}
	}else{
		cfg.SegmentSize = 5000
	}

// All other fields...
}
```
Not undoable, but probably not worth it in this case. It also means the "default" values would only get applied when coming in from JSON, and not when something directly calls NewColorDetector() either, which could cause issues. For now, I'm just going with "zero not allowed"